Use JOptionPane instead of JDialog

danielb987 opened the pull request:
Fix issue #6338

dheap commented:
> @rhwood Also "Start with new profile" should probably be "Change profile" and lead to the profile selector, which has a "New profile" button on it.

I think that's an excellent idea. "New Profile" can cause average user problems because it more often than not leads to apparent loss of roster (and all settings).

dheap commented:
"Restart Program" doesn't necessarily display the Profile Selector, unless the user has enabled that option.

danielb987 commented:
> > @rhwood Also "Start with new profile" should probably be "Change profile" and lead to the profile selector, which has a "New profile" button on it.
> 
> I think that's an excellent idea. "New Profile" can cause average user problems because it more often than not leads to apparent loss of roster (and all settings).

As I wrote in  #6338, "Change profile" will need more work (adding a new dialog) so I would like to get this PR merged first before doing that in case I mess things up.

rhwood commented:
The Change Profile dialog already exists—its the first thing openned if you have more than one profile.

danielb987 commented:
> The Change Profile dialog already exists—its the first thing openned if you have more than one profile.

Thanks. I found it.

`ProfileManager.saveActiveProfile()` throws an IOException. `JmriConfigurationManager.load()` throws an JmriException but not an IOException. What is the proper way to hande the exception in this case? Throw an JmriException with the IOException or handle the exception in a try-catch and log the error?

danielb987 commented:
`JmriConfigurationManager.load()` is the method that shows the error dialog that gives the option to edit connections if the connection cannot be opened.

I have got working code now:

```
if (Bundle.getMessage("ErrorDialogButtonChangeProfile").equals(selectedValue)) {
    ProfileManager manager = ProfileManager.getDefault();
    Profile last = manager.getActiveProfile();
    ProfileManagerDialog pmd = new ProfileManagerDialog(null, true, true);
    pmd.setLocationRelativeTo(null);
    pmd.setVisible(true);
    if (last == null || !last.equals(manager.getActiveProfile())) {
        try {
            manager.saveActiveProfile();
        } catch (IOException e) {
            log.error("Cannot save profile", e);
        }
    }
    // Restart program
    AppsBase.handleRestart();
}
```

rhwood commented:
This shows me the profile selector twice, but without changing how JMRI applications launch, there's nothing to do about that.

Some notes from the screen shot:
<img width="778" alt="screen shot 2018-12-22 at 07 32 08" src="https://user-images.githubusercontent.com/297232/50374429-5a63ef00-05bc-11e9-9043-9f3fdf11827c.png">
- How do I get the logs (text marked through in red)? We need to provide a button for that or remove that text.
- We know the program name (see title in green), so can we use it in the buttons (circled in green)?
- The language for the halt program ("Exit") button is different than the language for halting the program on a successful launch ("Quit"). These should be consistent (I'd use Quit, since JMRI applications have used that for a long time).
- If you want to use a button with options, use the [JSplitButton](http://rhwood.github.io/jsplitbutton/) (JMRI includes it as a library).
- We may want to include a line under the list of errors: "If you need to plug in or turn on anything, do so before restarting <app name>."
- What is the value of offering to shutdown/quit/exit as well as restarting the app with a button (even without a button Alt+F4 / Ctrl-Q / Cmd-Q can be used to exit the app)?

rhwood commented:
I hit send too soon: "Copy" copies the selection. Can it be disabled without a selection, and can a "Copy All" menu item be present to copy all items in the list, regardless of selection?

danielb987 commented:
> * How do I get the logs (text marked through in red)? We need to provide a button for that or remove that text.

Text is removed.

> * We know the program name (see title in green), so can we use it in the buttons (circled in green)?

Fixed.

> * The language for the halt program ("Exit") button is different than the language for halting the program on a successful launch ("Quit").

Fixed.

> * If you want to use a button with options, use the [JSplitButton](http://rhwood.github.io/jsplitbutton/) (JMRI includes it as a library).

Should I use it here? If so, where and why?

> * We may want to include a line under the list of errors: "If you need to plug in or turn on anything, do so before restarting ."

Fixed.

> * What is the value of offering to shutdown/quit/exit as well as restarting the app with a button (even without a button Alt+F4 / Ctrl-Q / Cmd-Q can be used to exit the app)?

The user interface should not require that the user knows the keyboard shortcuts. A new user might not know or remember these keyboard short cuts. Restart was added due to #5849.

danielb987 commented:
> I hit send too soon: "Copy" copies the selection. Can it be disabled without a selection, and can a "Copy All" menu item be present to copy all items in the list, regardless of selection?

Fixed.

CopyAll may result in a list with some html and some non-html. Example:
```
Unable to create connection "LocoNet" (L).
<html>&nbsp;&nbsp;&nbsp;&nbsp;Serial port COM4 not found.</html>Serial port COM4 not found.
Unable to run startup actions due to earlier failures.
```

danielb987 commented:
Added space to restart Travis after:
```
[ERROR] Failures: 
[ERROR]   LnOpsModeProgrammerTest.testBoardWriteTimeout:288 Correct thread
```

rhwood commented:
The Apple, Microsoft, and GNOME (Linux), Human Interface Guidelines all state there should be no more than 3 buttons in a modal dialog like this unless the forth button is a help button. These guidelines are in place because too many choices, especially when things aren't working, can lead to cognitive overload and user frustration.

Therefor, my question about including a quit button and a restart button is "why both?" Restart I get: I've had to take an external action to fix the connection (plug something in or turn something on), and restarting the application completes the fix; quit (given the option of changing profiles) I don't get--are we expecting that some connections will be a completely unrecoverable error from which the user might as well give up?

If both options are needed, then a split button that shows "Restart", but allows the user to select "Quit" from the menu needs to be displayed.

Also, why am I not allowed to continue using this application despite the error?

danielb987 commented:
> * If you want to use a button with options, use the [JSplitButton](http://rhwood.github.io/jsplitbutton/) (JMRI includes it as a library).

Since I changed the code to use JOptionPane instead of JDialog, I cannot use JSplitButton because I don't create the buttons myself.

```
Object[] options = new Object[] {
    Bundle.getMessage("ErrorDialogButtonQuitProgram", Application.getApplicationName()),
    Bundle.getMessage("ErrorDialogButtonRestartProgram", Application.getApplicationName()),
    Bundle.getMessage("ErrorDialogButtonChangeProfile"),
    Bundle.getMessage("ErrorDialogButtonEditConnections")
};

...

JOptionPane pane = new JOptionPane(
        new Object[] {
            (list instanceof JList) ? Bundle.getMessage("InitExMessageListHeader") : null,
            list,
            "<html><br></html>", // Add a visual break between list of errors and notes // NOI18N
            Bundle.getMessage("InitExMessageLogs"), // NOI18N
            Bundle.getMessage("ErrorDialogConnectLayout"), // NOI18N
        },
        JOptionPane.ERROR_MESSAGE,
        JOptionPane.DEFAULT_OPTION,
        null,
        options
);
```

danielb987 commented:
I agree with @bobjacobsen. If you click on "Quit" by misstake, it's easy to start the program again. For a user that wants to quit but is unsure how the SplitButton works, it's easy to miss the "Quit" option. So I have switched them to have Quit on the button and Restart as an option.

bobjacobsen commented:
> If both options are needed, then a split button that shows "Restart", but allows the user to select "Quit" from the menu needs to be displayed.

Could you say a few works on why you see this as more user friendly than showing two buttons?  It looks more confusing to me, but perhaps I’m not looking at it right.

danielb987 commented:
@rhwood : Can you answer this question from @bobjacobsen ? So I know if this PR is complete or if I should change it.

> If both options are needed, then a split button that shows "Restart", but allows the user to select "Quit" from the menu needs to be displayed.
> Could you say a few works on why you see this as more user friendly than showing two buttons? It looks more confusing to me, but perhaps I’m not looking at it right.

rhwood commented:
I’m working on a complete answer to his question—it’s complex and is requiring thought.

rhwood commented:
The various Human Interface Guides (HIGs) for the operating systems that JMRI supports all stipulate that a user should be warned of a dangerous or erroneous condition with a dialog containing not more than three buttons. Each button represents a path out of that dangerous or erroneous state. I am assuming, at least on the part of Apple and Microsoft, that the HIGs cap the number of buttons based on some sound research concerning people's ability to deal with choices (especially when not given access to a complete discussion of what the choices mean) and the number of choices at which too many options likely results in an unwanted outcome. It seems to me that the three paths are _continue_ (accepting the erroneous condition or taking the dangerous action), _abort_ (back out of the erroneous condition or not take the dangerous action), or _take corrective action_ (allow the user to fix the erroneous condition or modify the action to not be dangerous).

The most trivial example of the above is a warning about overwriting a file when saving a new file, with the options to _continue_ (overwrite the file, loosing whatever was in it previously), _abort_ (don't save anything), or _take corrective action_ (change the name or location of the file being saved).

When JMRI is at risk of not functioning correctly, mangling a connection profile, or even damaging a layout, because a connection is not working (due to either opening the wrong profile or failing to plug something in or power something on) that is a dangerous condition that may be erroneous.

The problem here is more complex than the overwriting a file example, but does not have the same three paths. Instead, this dialog provides four paths, one of which is _abort_ ("quit"), and two of which are _take corrective action_ ("change profile and then restart" or "edit connection preferences and then either restart or quit"), and one of which is _possibly repeat the same problem_ ("restart").

I understand that the _possibly repeat the same problem_ option was explicitly requested, but I don't think it should be obvious, which is why I would recommend that it be available, but not obvious.

I think that a better, but significantly more complex to implement set of choices would be to allow the user from three choices to:
- plugin or power on connections, dump all items from the InstanceManager, and reload the same configuration (aka _continue_),
- open limited preferences showing just the connections and profiles panes, and restart with edited connection or profiles (aka _take corrective action_),
- quit the application (aka _abort_) (and if they really need it, find the obscured restart command)

danielb987 commented:
Thanks for your answer.

> * plugin or power on connections, dump all items from the InstanceManager, and reload the same configuration (aka _continue_),

I don't see why this would be better than giving the user the ability to restart the program. For example, JMRI restarts the program when the user has changed the preferences. In what way would it be better to clear the InstanceManager and reload the configuration? Would the user see the difference?

> * open limited preferences showing just the connections and profiles panes, and restart with edited connection or profiles (aka _take corrective action_),

I think this could be even more confusing to users. _Change profile_ and _Edit connections_ are two complete different tasks and I personally thinks that open "Preferences" with only these two would confuse at least myself a lot.

I other words: I'm confused by the fact that _Config profiles_ is on Preferences in the first place. I don't think it belongs there at all, and if it should be there, it should be the first item in the list. The problem is that _Config profiles_ handles other profiles than the current profile, but everything else on Preferences handles the current profile.

For me, it would be better if _Config profiles_ was moved out from the Preferences dialog and had it's own menu choice _Edit_ / _Config profiles_ .

KenC57 commented:
I did a read, and then read again a few hours later, Randall, I think you
reversed some terms from what most would expect. You said:
----
I think that a better, but significantly more complex to implement set of
choices would be to allow the user from three choices to:
.	plugin or power on connections, dump all items from the
InstanceManager, and reload the same configuration (aka continue),
.	open limited preferences showing just the connections and profiles
panes, and restart with edited connection or profiles (aka take corrective
action),
.	quit the application (aka abort) (and if they really need it, find
the obscured restart command)
----
I think the first item is the 'abort/retry', it expects the problem is
external to JMRI. Since nothing is really working at this point, doing the
restart is the clean way since you are not expecting a change internally.

Second item, yes that is a internally corrective action.

The third item is really the same case as the first, you are doing nothing
inside of JMRI at this time, and this brings it back to really two cases. 

But do we want to allow a forth (third?) case of continue with the broken
configuration? Of Randall's three choices, it would involve editing the
profile/picking a profile that does match the hardware at hand. At least
that's how I'm reading it.

bobjacobsen commented:
All of this started as an attempt to make it easier for people to deal with their hardware-configuration mistakes, i.e. forgetting to plug in a cable.  There's is _definitely_ a point of diminishing returns on things like this.

The most basic need is to describe the problem, preferably in detail, and then provide just two options:
- Quit.  If you need to fix cables, etc, just quit and manually relaunch the program when you're done.
- Continue. You want to continue, i.e. to debug the problem, change the preferences to new values, or whatever.

@danielb987 has added a valuable 3rd option, which I'd slightly modify to:
- Edit connection preferences for the connection with the error, then save&quit or just quit.  You can launch again if desired.  This is a little bit less confusing than using option 2 "continue", because you'll be focused on the connection preferences.  So perhaps it's an OK bit of help for the user. But I think it's already at or beyond the point of diminishing return, and I wouldn't go any further in that direction.

I do _not_ favor a "restart" option, hidden or otherwise.  It saves negligible time, and it's clearly confusing.

rhwood commented:
I'm in agreement with @bobjacobsen on this. I *need* to be able to open a profile that can't make its connections to be able to ensure that a roster issue I am encountering an otherwise working profile (when hardware is connected) is or is not present in trunk, and this damned dialog barring that is completely blocking even looking into the issue I am encountering.

danielb987 commented:
I have changed the dialog to this:
![jmri](https://user-images.githubusercontent.com/20255317/50563172-dd9ee680-0d1a-11e9-8ef3-b3d05f4926a5.png)

Continue results in opening PanelPro and showing the Preference window.

I'm not sure what to do with the "Edit connections". Remove it or keep it? Should I change it and if so, how? The "save" button in the Preference window restarts the program and I personally thinks it would be confusing if the "Edit connections" dialog had a "Save" button which quit the program.

bobjacobsen commented:
> I'm not sure what to do with the "Edit connections". Remove it or keep it? Should I change it and if so, how?

I think it’s fine as is. 

> The "save" button in the Preference window restarts the program and I personally thinks it would be confusing if the "Edit connections" dialog had a "Save" button which quit the program.

IIRC, save in the main preferences gives you the option to restart or not. 

The edit-preferences from edit-connections here can’t do that, so maybe there should just be a Save (that just saves) and a Quit button.

danielb987 commented:
> IIRC, save in the main preferences gives you the option to restart or not.

Yes.

> The edit-preferences from edit-connections here can’t do that, so maybe there should just be a Save (that just saves) and a Quit button.

I can do that. But if nothing happens when the user clicks "Save", it may confuse the user to think that nothing has happen. Maybe I should put up some dialog after I have saved the data?

danielb987 commented:
Travis timed out. AppVeyor failed due to:
```
Testcase: testGetAndSetDisplayLevel(jmri.jmrit.display.LinkingLabelTest):	FAILED
Unexpected ERROR or higher messages emitted
junit.framework.AssertionFailedError: Unexpected ERROR or higher messages emitted
    at jmri.util.JUnitUtil.tearDown(JUnitUtil.java:270)
    at jmri.jmrit.display.LinkingLabelTest.tearDown(LinkingLabelTest.java:89)
```

rhwood commented:
Every PreferencesPanel has a method to tell if restarting is required ( http://jmri.org/JavaDoc/doc/jmri/jmrix/swing/ConnectionsPreferencesPanel.html#isRestartRequired-- ). Why can’t you use it to make the determination just as the standard Prefrrences window does?

danielb987 commented:
> Every PreferencesPanel has a method to tell if restarting is required ( http://jmri.org/JavaDoc/doc/jmri/jmrix/swing/ConnectionsPreferencesPanel.html#isRestartRequired-- ). Why can’t you use it to make the determination just as the standard Prefrrences window does?

What should happen if the method say that restart is not required? Continue to start the program? Quit the program?

danielb987 commented:
@bobjacobsen@rhwood

danielb987 commented:
@bobjacobsen
@rhwood
Should I change the PreferenceDialog? If not, I think this PR is ready to be merged, except that Travis and AppVeyor needs to be restarted.

> > Every PreferencesPanel has a method to tell if restarting is required ( http://jmri.org/JavaDoc/doc/jmri/jmrix/swing/ConnectionsPreferencesPanel.html#isRestartRequired-- ). Why can’t you use it to make the determination just as the standard Prefrrences window does?
> 
> What should happen if the method say that restart is not required? Continue to start the program? Quit the program?

bobjacobsen commented:
Restarted the CI

rhwood commented:
I'm fine with the current state of this.

bobjacobsen commented:
Thanks for all the work on this, everybody.  All CI has passed now, so only the WIP label is holding it back.  With @danielb987 and @rhwood comments above, Im going to remove that and merge this.  

Thanks again to all!
